src/main/java/com/minecolonies/coremod/inventory/CraftingGUIBuilding.java

```diff
@@ -0,0 +1,350 @@
+        }
+
+        for (i = 0; i < PLAYER_INVENTORY_COLUMNS; i++)
+        {
+            addSlotToContainer(new Slot(
+                    playerInventory, i,
+                    PLAYER_INVENTORY_INITIAL_X_OFFSET + i * INVENTORY_OFFSET_EACH,
+                    PLAYER_INVENTORY_HOTBAR_OFFSET
+            ));
+        }
+
+        this.onCraftMatrixChanged(this.craftMatrix);
+    }
+
+    @Override
+    protected final Slot addSlotToContainer(final Slot slotToAdd)
+    {
+        return super.addSlotToContainer(slotToAdd);
+    }
+
+    /**
+     * Callback for when the crafting matrix is changed.
+     */
+    @Override
+    public void onCraftMatrixChanged(final IInventory inventoryIn)
+    {
+        if (!worldObj.isRemote)
+        {
+            EntityPlayerMP entityplayermp = (EntityPlayerMP) player;
+            ItemStack itemstack = ItemStack.EMPTY;
+            IRecipe irecipe = CraftingManager.findMatchingRecipe(craftMatrix, worldObj);
+            if (irecipe != null && (irecipe.isDynamic()
+                    || !worldObj.getGameRules().getBoolean("doLimitedCrafting")
+                    || entityplayermp.getRecipeBook().isUnlocked(irecipe)
+                    || entityplayermp.isCreative()))
+            {
+                this.craftResult.setRecipeUsed(irecipe);
+                itemstack = irecipe.getCraftingResult(this.craftMatrix);
+            }
+
+            this.craftResult.setInventorySlotContents(0, itemstack);
+            entityplayermp.connection.sendPacket(new SPacketSetSlot(this.windowId, 0, itemstack));
+        }
+
+        super.onCraftMatrixChanged(inventoryIn);
+    }
+
+    @Override
+    public boolean canInteractWith(final EntityPlayer playerIn)
+    {
+        return true;
+    }
+
+    @Override
+    public ItemStack slotClick(int slotId, int clickedButton, ClickType mode, EntityPlayer playerIn)
+    {
+        if (slotId >= 1 && slotId < 5)
+        {
+            // 1 is shift-click
+            if (mode == ClickType.PICKUP
+                    || mode == ClickType.PICKUP_ALL
+                    || mode == ClickType.SWAP)
+            {
+                Slot slot = this.inventorySlots.get(slotId);
+
+                ItemStack dropping = playerIn.inventory.getItemStack();
+
+                return handleSlotClick(slot, dropping);
+            }
+
+            return ItemStack.EMPTY;
+        }
+
+        return super.slotClick(slotId, clickedButton, mode, playerIn);
+    }
+
+    public ItemStack handleSlotClick(Slot slot, ItemStack stack)
+    {
+        if (stack.getCount() > 0)
+        {
+            ItemStack copy = stack.copy();
+            copy.setCount(1);
+            slot.putStack(copy);
+        }
+        else if (slot.getStack().getCount() > 0)
+        {
+            slot.putStack(ItemStack.EMPTY);
+        }
+
+        return slot.getStack().copy();
+    }
+
+    @Nullable
+    @Override
+    public ItemStack transferStackInSlot(final EntityPlayer playerIn, final int index)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 7, down to the maximum allowed 6. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)